Add Makefile for building and uploading the Spin code

dgelessus opened the pull request:
As an alternative to the workflow based on the Propeller IDE, for users who cannot use the Propeller IDE for some reason (e. g. I couldn't find an arm64 build of the Propeller IDE) or who simply prefer the command line.

Tested on Linux (Debian 12, Linux 6.1.0-32-arm64), using Parallax [OpenSpin](https://github.com/parallaxinc/OpenSpin) and [PropLoader](https://github.com/parallaxinc/PropLoader) built from source.

dgelessus committed:
Add Makefile for building and uploading the Spin code

dgelessus committed:
Add missing proploader flag to actually program the EEPROM

Without -e, it just loads the new firmware into RAM and it will be lost
after a power cycle.

dgelessus commented:
For some reason, the `proploader` call always shows me an error at the end, but the firmware seems to be written correctly anyway...

```
Downloading file to port /dev/ttyUSB0
30456 bytes sent
Verifying RAM
Programming EEPROM
ERROR: EEPROM verify failed
ERROR: Download failed
```

dgelessus committed:
Add comments

fjmilens3 commented:
I had to edit the `Makefile` to add my port via `-p` but I get no errors at all:

```
% make upload
proploader -e -s -p /dev/cu.usbserial-AK04L68D cody_computer.binary
Opening file 'cody_computer.binary'
Downloading file to port /dev/cu.usbserial-AK04L68D
30456 bytes sent                  
Verifying RAM
Programming EEPROM
Download successful!
```

That's on a 2019 MacBook Pro (Intel) / MacOS Ventura 13.6.1, not sure about the revision of my Prop Plug but I've had it for almost a decade.

dgelessus committed:
Add reset flag to proploader call

This makes the Cody Computer automatically boot after the firmware is
uploaded (default behavior is to leave the Propeller powered off,
apparently) and for some reason also fixes the "EEPROM verify failed"
errors I get at the end of the upload.

dgelessus committed:
Add convenience variable for specifying serial port for upload

dgelessus commented:
I've added the `-r` flag as discussed and this indeed resolves the "EEPROM verify failed" error for me.

Also added a convenience variable for configuring the serial port for the upload. Now you can run e. g. `make PORT=/dev/cu.usbserial-AK04L68D upload` to upload through your serial port.

fjmilens3 commented:
I'll test it out locally on my Mac, if it works I'll merge it in (I can't imagine there being any problems).

fjmilens3 commented:
Quick tests on my Mac, looks good.

Test 1: `-P` to list ports (expected failure because correct port isn't provided, but lists ports)

```
frederick@Fredericks-MacBook-Pro Spin % PROPLOADER_FLAGS='-P' make upload 
proploader -P -e -r -s cody_computer.binary
/dev/cu.BLTH
/dev/cu.usbserial-AK04L68D
Opening file 'cody_computer.binary'
ERROR: Unable to connect to port /dev/cu.BLTH
make: *** [upload] Error 1
```

Test 2: `-p` to specify port.

```
frederick@Fredericks-MacBook-Pro Spin % PROPLOADER_FLAGS='-p/dev/cu.usbserial-AK04L68D' make upload
proploader -p/dev/cu.usbserial-AK04L68D -e -r -s cody_computer.binary
Opening file 'cody_computer.binary'
Downloading file to port /dev/cu.usbserial-AK04L68D
30456 bytes sent                  
Verifying RAM
Programming EEPROM
Download successful!
```

Test 3: Builds work (deleted all .bin outputs before running)

```
frederick@Fredericks-MacBook-Pro Spin % rm codybasic.bin
frederick@Fredericks-MacBook-Pro Spin % rm cody_computer.binary
frederick@Fredericks-MacBook-Pro Spin % rm ../Tass/codybasic.bin
frederick@Fredericks-MacBook-Pro Spin % make                    
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C ../Tass codybasic.bin
64tass --mw65c02 --nostart -o codybasic.bin codybasic.asm
64tass Turbo Assembler Macro V1.59.3120
64TASS comes with ABSOLUTELY NO WARRANTY; This is free software, and you
are welcome to redistribute it under certain conditions; See LICENSE!

Assembling file:   codybasic.asm
Output file:       codybasic.bin
Data:       8025   $e000-$ff58   $1f59
Gap:          71   $ff59-$ff9f   $0047
Data:         79   $ffa0-$ffee   $004f
Gap:          13   $ffef-$fffb   $000d
Data:          4   $fffc-$ffff   $0004
Passes:            3
cp ../Tass/codybasic.bin codybasic.bin
openspin -o cody_computer.binary cody_computer.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2018 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.81 Compiled on Apr 12 2025 22:16:04
Compiling...
cody_computer.spin
|-cody_video.spin
  |-cody_line.spin
|-cody_audio.spin
|-cody_uart.spin
Done.
Program size is 30456 bytes
```

fjmilens3 commented:
Thank you, @dgelessus!
